packages/simulation/src/convert.ts

```diff
@@ -0,0 +1,647 @@
+  const numberOfNFAStates: number = nfaGraphStates.length
+  let doesFinalExist: boolean = false
+  for (let curElem = 0; curElem < numberOfNFAStates; curElem++) {
+    if (nfaGraphStates[curElem].isFinal) {
+      doesFinalExist = true
+    }
+  }
+  if (!doesFinalExist) {
+    return false
+  } else {
+    return true
+  }
+}
+
+// This will check to ensure that the graph passed in has an initial state before continuing
+export const initialStateIsPresent = (nfaGraphInitialState: StateID): boolean => {
+  if (nfaGraphInitialState == null) {
+    return false
+  } else {
+    return true
+  }
+}
+
+// This will check if the merged state is different from the current state's merged state
+function arraysEqual (arr1, arr2) {
+  if (arr1.length !== arr2.length) {
+    return false
+  }
+  for (let i = 0; i < arr1.length; i++) {
+    if (arr1[i] !== arr2[i]) {
+      return false
+    }
+  }
+  return true
+}
+
+// This will check that a given states transitions contain all the symbols
+function hasAllSymbols (transitions, symbols) {
+  for (const symbol of symbols) {
+    let symbolFound = false
+    for (const [, readSymbol] of transitions) {
+      if (readSymbol === symbol) {
+        symbolFound = true
+        break
+      }
+    }
+    if (!symbolFound) {
+      return false
+    }
+  }
+  return true
+}
+
+// This function will handle the logic for combining states when necessary
+function combineStates (initialTransitionTable: {[key: StateID]: [StateID, ReadSymbol][]}, symbolsArray: ReadSymbol[], curTransitionElem: number, symbolToStatesMap: {[key: ReadSymbol]: [StateID, StateID][]}, symbol: ReadSymbol): [string, { symbol: ReadSymbol, tostates: StateID[] }[]] {
+  const filteredTransitions = symbolToStatesMap[symbol].filter(trans => trans[0] === curTransitionElem)
+  if (filteredTransitions.length <= 1) {
+    return [null, null]
+  }
+
+  filteredTransitions.sort((a, b) => a[1] - b[1])
+
+  const curKeyCombination = filteredTransitions.map(trans => trans[1]).join(',')
+  const curStateCombination = curKeyCombination.split(',').map(Number)
+
+  const combinedState = symbolsArray.map(symbol => {
+    const tostates = curStateCombination.flatMap(curState => {
+      return initialTransitionTable[curState]
+        .filter(trans => trans[1] === symbol)
+        .map(trans => trans[0])
+    }).sort()
+
+    return { symbol, tostates }
+  })
+
+  return [curKeyCombination, combinedState]
+}
+
+// This will mimic STEP 1 of the procedure by ensuring that for every DFA state, there are no lambda transitions and the states are merged correctly
+export function removeLambdaTransitions (initialTransitionTable: {[key: StateID]: [StateID, ReadSymbol][]}, mergedStates: {[key: string]: StateID[]}, numberOfNFAStates: StateID, curInitialState: StateID, curFinalStates: StateID[], tempTransitionTable: {[key: StateID]: [StateID, ReadSymbol][]}): [{[key: StateID]: [StateID, ReadSymbol][]}, StateID, StateID[]] {
```

Comment:
Same here, these types seem to be repeated in a lot of places so best to move them into actual types